plutus-core/executables/pir/Main.hs

```diff
@@ -1,106 +1,288 @@
-type PIRError = PIR.Error PLC.DefaultUni PLC.DefaultFun (PIR.Provenance ())
-type PIRCompilationCtx a = PIR.CompilationCtx PLC.DefaultUni PLC.DefaultFun a
-data Command = Analyse IOSpec
-             | Compile COpts
-             | Print PrintOptions
-
-data COpts = COpts
-  { cIn       :: Input
-  , cOptimize :: Bool
-  }
-
-pCOpts :: Parser COpts
-pCOpts = COpts
-            <$> input
-            <*> switch' ( long "dont-optimize"
-                        <> help "Don't optimize"
-                        )
-  where
-    switch' :: Mod FlagFields Bool -> Parser Bool
-    switch' = fmap not . switch
-
-pPirOpts :: Parser Command
-pPirOpts = hsubparser $
-    command "analyse"
-        (info (Analyse <$> ioSpec) $
-            progDesc $
-              "Given a PIR program in flat format, deserialise and analyse the program, " <>
-              "looking for variables with the largest retained size.")
-  <> command "compile"
-        (info (Compile <$> pCOpts) $
-            progDesc $
-              "Given a PIR program in flat format, deserialise it, " <>
-              "and test if it can be successfully compiled to PLC.")
-  <> command "print"
-        (info (Print <$> printOpts) $
-            progDesc $
-              "Given a PIR program in flat format, " <>
-              "deserialise it and print it out textually.")
-
--- | Load flat pir and deserialize it
-loadPir :: Input -> IO (PirProg ())
-loadPir = loadASTfromFlat Named
-
-compile :: COpts -> PirProg () -> Either PIRError PLCTerm
-compile opts (PIR.Program _ pirT) = do
+
+type PIRErrorWithProvenance = PIR.Error PLC.DefaultUni PLC.DefaultFun (PIR.Provenance ())
+
+---------------- Types for command line options ----------------
+
+-- | A specialised format type for PIR. We don't support deBruijn or named deBruijn for PIR.
+data PirFormat = TextualPir | FlatNamed
+instance Show PirFormat
+    where show = \case { TextualPir  -> "textual"; FlatNamed -> "flat-named" }
+
+data PirOptimiseOptions = PirOptimiseOptions Input PirFormat Output PirFormat PrintMode
+
+data AnalyseOptions = AnalyseOptions Input PirFormat Output -- Input is a program, output is text
+
+data Language = PLC | UPLC
+instance Show Language
+    where show = \case { PLC  -> "plc"; UPLC -> "uplc" }
+
+-- | Compilation options: target language, whether to optimise or not, input and output streams and types
+data CompileOptions =
+    CompileOptions Language
+                   Bool   -- Optimise or not?
+                   Bool   -- True -> just report if compilation was successful; False -> write output
+                   Input
+                   PirFormat
+                   Output
+                   Format
+                   PrintMode
+
+data Command = Analyse  AnalyseOptions
+             | Compile  CompileOptions
+             | Convert  ConvertOptions
+             | Optimise PirOptimiseOptions
+             | Print    PrintOptions
+
+
+---------------- Option parsers ----------------
```

Comment:
Maybe I should put all of these in a separate file to improve readability.  I'd have to move the types too if I did that though.